src/object_based/great_circle_arc_intersection.hpp

```diff
@@ -0,0 +1,149 @@
+#pragma once
+#include <Eigen/Geometry>
+#include <boost/math/constants/constants.hpp>
+#include <limits>
+#include <utility>
+
+namespace ear {
+  /// tests for intersections between a great circle and an arc on the surface
+  /// of a unit sphere
+  class GreatCircleArcIntersection {
+    using Vec2 = Eigen::Vector2d;
+
+   public:
+    /// the plane of the arc is defined by front and right, which must be
+    /// orthogonal
+    ///
+    /// opening_angle is the angle between the centre of the arc, the centre of
+    /// the sphere, and the arc itself
+    ///
+    /// wrap angle is the angle between the mid-point of the arc, the centre of
+    /// the arc, and one end of the arc
+    ///
+    /// the vector from the centre of the arc to the mid-point is the same as
+    /// front
+    ///
+    /// if opening_angle is less than PI/2, the vector from the centre of the
+    /// sphere to the centre of the arc is cross(right, front)
+    GreatCircleArcIntersection(const Eigen::Ref<const Eigen::Vector3d> &front,
+                               Eigen::Vector3d right, double opening_angle,
+                               double wrap_angle) {
+      const double PI = boost::math::constants::pi<double>();
+      assert(std::abs(front.dot(right)) < 1e-6);
+
+      // for wide opening angles, invert so that the arc is always above the
+      // front-right plane
+      if (opening_angle > PI / 2) {
+        right *= -1;
+        opening_angle = PI - opening_angle;
+      }
+
+      height = cos(opening_angle);
+
+      Eigen::Vector3d normal = right.cross(front);
+      normal.normalize();
+
+      T.row(0) = -right;
+      T.row(1) = front;
+      T.row(2) = -normal;
+      r = std::abs(sin(opening_angle));
+      wrap_angle = std::min(std::abs(wrap_angle), PI);
+      double test_angle = wrap_angle - PI / 2;
+      test_vec = {sin(test_angle), cos(test_angle)};
+    }
+
+    /// does the great circle between p1_in and p2_in intersect the arc defined
+    /// in the constructor? p1_in and p2_in must not be equal.
+    bool intersects(const Eigen::Ref<const Eigen::Vector3d> &p1_in,
+                    const Eigen::Ref<const Eigen::Vector3d> &p2_in) const {
+      assert((p1_in - p2_in).squaredNorm() > 1e-6);
```

Comment:
In general I'd suggest breaking an algorithm with comments describing sections into sensibly named functions for each section that don't need comments. Can make it easier to get a high level overview of what's going on (also if you wind up needing a million parameters to make that work it's a smell that you maybe could do with a struct somewhere). However, with mathematical stuff it sometimes doesn't help much as it can be really abstract, so will leave it up to you as something to consider.